apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryCommand.Data.cs

```diff
@@ -15,62 +15,240 @@
 using System;
 using System.Data;
 using System.Data.Common;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax;
+using Google.Apis.Bigquery.v2.Data;
 
 namespace Google.Cloud.BigQuery.V2
 {
+    /// <summary>
+    /// Represents a command to execute on BigQuery.
+    /// This can currently only be a SQL statement (query or DML).
+    /// Other calls (such as <see cref="BigQueryConnection.ListRowsAsync(string,TableSchema,ListRowsOptions)"/>
+    /// are made directly on <see cref="BigQueryConnection"/>.
+    /// </summary>
     public sealed partial class BigQueryCommand : DbCommand
     {
+        /// <summary>
+        ///     CommandOptions provides options and what sort of command is being executed.
+        ///     It can never be null and defaults to a sql command (DML or Query).
+        /// </summary>
+        private BigQueryCommandOptions _commandOptions = new SqlCommandOptions();
+
+        internal CancellationTokenSource SynchronousCancellationTokenSource { get; } = new CancellationTokenSource();
+
         /// <inheritdoc />
-        public override void Cancel()
+        public override string CommandText
         {
-            throw new NotImplementedException();
+            get => CommandOptions.CommandText;
+            set => CommandOptions.CommandText = value;
         }
 
         /// <inheritdoc />
-        public override int ExecuteNonQuery()
+        public override int CommandTimeout
         {
-            throw new NotImplementedException();
+            get => CommandOptions.CommandTimeout.GetValueOrDefault((int) GetQueryResultsOptions.DefaultTimeout
+                .TotalSeconds);
+            set => CommandOptions.CommandTimeout = value;
         }
 
         /// <inheritdoc />
-        public override object ExecuteScalar()
+        public override CommandType CommandType
         {
-            throw new NotImplementedException();
+            get => CommandType.Text;
+            set
+            {
+                if (value != CommandType.Text)
+                    throw new NotSupportedException("Cloud BigQuery only supports CommandType.Text.");
+            }
         }
 
+        /// <summary>
+        /// Represents the type of Command.  Currently the only supported value is
+        /// <see cref="BigQueryCommandType.Sql"/>.
+        /// </summary>
+        public BigQueryCommandType BigQueryCommandType => CommandOptions.BigQueryCommandType;
+
         /// <inheritdoc />
-        public override void Prepare()
+        public override UpdateRowSource UpdatedRowSource
+        {
+            get => UpdateRowSource.None;
+            set
+            {
+                if (value != UpdateRowSource.None)
+                    throw new NotSupportedException(
+                        "Cloud BigQuery does not support updating datasets on update/insert queries."
+                        + " Please use UUIDs instead of auto increment columns, which can be created on the client.");
```

Comment:
this is most often used in the scenario where you have an autoincrement column in the database and you need to set the updaterowsource to "both" or "output" to have the cell in the local dataset display the correct data after the insert.  So the comment addresses the underlying use case advising the user to use UUIDs for key columns instead of autoincrement fields.